### PR TITLE
[Inference]Support FP16/BF16 Flash Attention 2 And Add high_precision Flag To Rotary Embedding

### PR DESCRIPTION
**Pull request by yuehuayingxueluo**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.
**pytest:**
image

**benchmark（kv memcopy + context flash attn，input len 128，output len 32）：**
| flash attn 2  | triton |  vllm  |
|  ----  | ----  | ----  |
| 0.229ms  | 0.323 ms | 0.271ms |

***flash attn 2(0.229ms)：***
image

***triton（0.323 ms）：***
image

***vllm（0.271ms）***
image

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

**Review by Courtesy-Xs — CHANGES_REQUESTED** [review]
Can you list a data sheet to make it clearer, eg
image

File: extensions/csrc/cuda/context_kv_cache_memcpy_kernel.cu [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+    int block_size = key_cache.size(2);
+    int batch_size = block_tables.size(0);
+
+    int64_t key_stride = key.stride(0);
+    int64_t value_stride = value.stride(0);
+    int block_table_stride = block_tables.stride(0);
+
+    int vec_size = get_vec_size<scalar_t>(key);
+
+    if (head_dim % vec_size != 0) {
```
Comment (Courtesy-Xs) [inline extensions/csrc/cuda/context_kv_cache_memcpy_kernel.cu:75]: Maybe some tail process can be added in the future so it is able to handle this situation

Comment (yuehuayingxueluo) [reply extensions/csrc/cuda/context_kv_cache_memcpy_kernel.cu:96]: Done

File: colossalai/inference/config.py [hunk 2]
```diff
@@ -55,7 +55,7 @@ class InferenceConfig:
         pp_size (int): Pipeline parallel size, defaults to 1.
         micro_batch_size (int): the micro batch size, defaults to 1. Only useful when `pp_size` > 1.
         micro_batch_buffer_size (int): the buffer size for micro batch. Normally, it should be the same as the number of pipeline stages.
-
+        high_precision(Optional[bool]): Whether to cast fp16 to fp32 for calculation, defaults to False.
```
Comment (Courtesy-Xs) [inline colossalai/inference/config.py:58]: cast from fp16 to fp32

Comment (yuehuayingxueluo) [reply colossalai/inference/config.py:58]: Is anything wrong?

Comment (yuehuayingxueluo) [reply colossalai/inference/config.py]: I have changed the way I phrase it now.

File: tests/test_infer/test_ops/cuda/test_rotary_embdding_unpad.py [hunk 3]
```diff
@@ -54,17 +55,26 @@ def test_rotary_emb(BATCH_SIZE, SEQ_LEN, H, D, dtype):
 
     kv_seq_lengths = past_kv_seq_lengths + 1
     block_tables = block_tables.to(device="cuda")
-    q_ref = torch_rotary_emb(new_q, cos[:BATCH_SIZE], sin[:BATCH_SIZE])
-    k_ref = torch_rotary_emb(new_k, cos[:BATCH_SIZE], sin[:BATCH_SIZE])
+
+    if high_precision:
```
[the remark this answers is not in the document]
Comment (yuehuayingxueluo) [reply tests/test_infer/test_ops/cuda/test_rotary_embdding_unpad.py]: Done

**Review by Courtesy-Xs — APPROVED** [review]
LGTM

File: colossalai/inference/config.py [hunk 4]
```diff
@@ -108,6 +109,9 @@ def _verify_config(self) -> None:
             self.dtype in _ALLOWED_DTYPES
         ), f"Expected dtype to be in {_ALLOWED_DTYPES} but found an unknown dtype: {self.dtype}"
 
+        if self.dtype == torch.float32:
+            self.high_precision = False
+
```
Comment (yuanheng-zhao) [inline colossalai/inference/config.py:114]: Might want to refine the comment of `high_precision`. It's sort of ambiguous here that setting `high_precision` as `False` for data type float32. Maybe you want to express "skip using casting when the data type is float32"?

Comment (yuehuayingxueluo) [reply colossalai/inference/config.py:114]: Yes.

Comment (yuehuayingxueluo) [reply colossalai/inference/config.py:114]: Enabling high_precision for data type float32 is meaningless.

File: colossalai/inference/modeling/models/nopadding_llama.py [hunk 5]
```diff
@@ -436,6 +481,5 @@ def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
         """
         hidden_states = hidden_states.expand(2, -1, -1)
         gate_up_proj_out = torch.bmm(hidden_states, self.gate_up_weight)
-        act_out = torch.nn.functional.silu(gate_up_proj_out[0], inplace=True)
-        tmp_out = act_out * gate_up_proj_out[1]
-        return torch.mm(tmp_out, self.down_proj_weight)
+        act_out = inference_ops.silu_and_mul(gate_up_proj_out)
```
Comment (yuanheng-zhao) [inline colossalai/inference/modeling/models/nopadding_llama.py:484]: Where does `use_cuda_kernel` go

Comment (yuehuayingxueluo) [reply colossalai/inference/modeling/models/nopadding_llama.py:484]: The silu_and_mul doesn't has triton kernel, so we may not need use_cuda_kernel.

File: extensions/csrc/common/micros.h [hunk 6]
```diff
@@ -56,6 +56,23 @@
...
+    }                                                                         \
+    case true: {                                                              \
+      const bool high_precision = true;                                       \
+      DISPATCH_FLOAT_HALF_AND_BFLOAT(TYPE, NAME, __VA_ARGS__);                \
+      break;                                                                  \
+    }                                                                         \
+    default:                                                                  \
+      AT_ERROR("HIGH_PRECISION must be bool, but get ", HIGH_PRECISION, "."); \
+  }
+
```
Comment (yuanheng-zhao) [inline extensions/csrc/common/micros.h:75]: This macro is redundant and non-functional as referring to your subsequent usage: 

```
    DISPATCH_FLOAT_HALF_AND_BFLOAT_WITH_HIGH_PRECISION(
        high_precision,
        query.scalar_type(),
        "rotary_embedding_and_cache_copy",
        apply_rotary_embedding_and_cache_copy<scalar_t>(
        apply_rotary_embedding_and_cache_copy<scalar_t, high_precision>(
            query,
            key,
            value,
            cos,
            sin,
            key_cache,
            value_cache,
            sequence_lengths,
            block_tables
        );)
```

Just use `DISPATCH_FLOAT_HALF_AND_BFLOAT`

File: extensions/csrc/cuda/fused_rotary_emb_and_cache_kernel.cu [hunk 7]
```diff
@@ -436,12 +441,14 @@ void rotary_embedding_and_cache_copy(
     at::Tensor& key_cache,           // [num_blocks, head_num, block_size, head_dim]
     at::Tensor& value_cache,         // [num_blocks, head_num, block_size, head_dim]
     at::Tensor& sequence_lengths,    // [batch_size]
-    at::Tensor& block_tables)        // [batch_size, max_seq_len]
+    at::Tensor& block_tables,        // [batch_size, max_seq_len]
+    bool high_precision)
 {
-    DISPATCH_FLOAT_HALF_AND_BFLOAT(
+    DISPATCH_FLOAT_HALF_AND_BFLOAT_WITH_HIGH_PRECISION(
```
Comment (yuanheng-zhao) [inline extensions/csrc/cuda/fused_rotary_emb_and_cache_kernel.cu:447]: Same as above

**Review by yuanheng-zhao — CHANGES_REQUESTED** [review]

**Review by yuanheng-zhao — APPROVED** [review]

**Pull request merged**